param/parameterized.py

```diff
@@ -1779,15 +1928,139 @@ def _setup_params(self_, **params):
             self_._instantiate_param(p, deepcopy=False)
 
         ## keyword arg setting
+        deps, refs = {}, {}
+        objects = self.param.objects(instance=False)
         for name, val in params.items():
             desc = self_.cls.get_param_descriptor(name)[0] # pylint: disable-msg=E1101
             if not desc:
                 raise TypeError(
                     f"{self.__class__.__name__}.__init__() got an unexpected "
                     f"keyword argument {name!r}"
                 )
-            # i.e. if not desc it's setting an attribute in __dict__, not a Parameter
-            setattr(self, name, val)
+
+            pobj = objects.get(name)
+            if pobj is None or not pobj.allow_refs:
+                # Until Parameter.allow_refs=True by default we have to
+                # speculatively evaluate a values to check whether they
+                # contain a reference and warn the user that the
+                # behavior may change in future.
+                if name not in self_.cls._param__private.explicit_no_refs:
+                    try:
+                        ref, _, resolved, _ = self_._resolve_ref(pobj, val)
+                    except Exception:
+                        ref = None
+                    if ref:
+                        warnings.warn(
+                            f"Parameter {name!r} on {pobj.owner} is being given a valid parameter "
+                            f"reference {val} but is implicitly allow_refs=False. "
+                            "In future allow_refs will be enabled by default and "
+                            f"the reference {val} will be resolved to its underlying "
+                            f"value {resolved}. Please explicitly set allow_ref on the "
+                            "Parameter definition to declare whether references "
+                            "should be resolved or not.",
+                            category=_ParamFutureWarning,
+                            stacklevel=2,
```

Comment:
Have you checked if the stack level is correct?

Reply:
Good catch, I haven't yet.

Reply:
Fixed.